third_party/terraform/utils/k8s_operation.go

```diff
@@ -0,0 +1,113 @@
+package google
+
+import (
+	"fmt"
+)
+
+// K8sOperation is a struct that can contain a Kubernetes resource's Status block. It is not
+// intended to be used for anything other than polling for the success of the given resource.
+type K8sOperation struct {
+	Metadata struct {
+		Name      string
+		Namespace string
+		SelfLink  string
+	}
+	Status struct {
+		Conditions []struct {
+			Type    string
+			Status  string
+			Reason  string
+			Message string
+		}
+	} `json:"status"`
+}
+
+// K8sOperationWaiter allows for polling against an arbitrary resource that implements the
+// Kubernetes status schema.
+type K8sOperationWaiter struct {
+	Config   *Config
+	Project  string
+	Location string
+	WaitURL  string
+	Op       K8sOperation
+}
+
+// State will return a string representing the status of the Ready condition.
+// No other conditions are currently returned as part of the state.
+func (w *K8sOperationWaiter) State() string {
+	for _, condition := range w.Op.Status.Conditions {
+		if condition.Type == "Ready" {
+			return fmt.Sprintf("%s:%s", condition.Type, condition.Status)
+		}
+	}
+	return "NotImplemented"
+}
+
+func (w *K8sOperationWaiter) Error() error {
+	if len(w.Op.Status.Conditions) == 0 {
+		// When initially created the returned object doesn't have a
+		// status block yet so continue polling until it does.
+		return nil
+	}
+
+	for _, condition := range w.Op.Status.Conditions {
+		if condition.Type == "Ready" && condition.Status == "False" {
+			return fmt.Errorf("%s - %s", condition.Reason, condition.Message)
+		}
+	}
+	return nil
+}
+
+func (w *K8sOperationWaiter) IsRetryable(error) bool {
+	return false
+}
+
+func (w *K8sOperationWaiter) OpName() string {
+	if w == nil {
+		return "waiter:<nil>"
+	}
+
+	return w.Op.Metadata.SelfLink
+}
+func (w *K8sOperationWaiter) PendingStates() []string {
+	return []string{"Ready:Unknown"}
+}
+func (w *K8sOperationWaiter) TargetStates() []string {
+	return []string{"Ready:True"}
```

Comment:
These kind of Pending / Target states should probably be in configuration and not templates.

Reply:
agreed - but this was my "1/2 custom" compromise where rather than fully templatize the operation I would handwrite the operation until it needed to be configured. FWIW I think all of these resources are going to be using ready:true, so until we see if this is a product/op level configuration or a resource level configuration I thought I'd leave it hard coded.